lahja/endpoint.py

```diff
@@ -489,28 +490,23 @@ def remove():
 
         self._queues[event_type].append(queue)
         self._notify_subscriptions_changed()
-        i = None if num_events is None else 0
-        while True:
-            try:
-                yield await queue.get()
-            except GeneratorExit:
-                self._queues[event_type].remove(queue)
-                self._notify_subscriptions_changed()
-                break
-            except asyncio.CancelledError:
-                self._queues[event_type].remove(queue)
-                self._notify_subscriptions_changed()
-                break
-            else:
-                if i is None:
-                    continue
 
-                i += 1
+        if num_events:
+            iterations = itertools.repeat(True, num_events)
+        else:
+            iterations = itertools.repeat(True)
 
-                if i >= cast(int, num_events):
-                    self._queues[event_type].remove(queue)
-                    self._notify_subscriptions_changed()
+        try:
+            for _ in iterations:
```

Comment:
Instead of the `itertools.repeat(True, num_events)` wouldn't something like `for i in range(num_events)` come more natural?

Reply:
Yeah, I'm pretty frustrated by this code! But `range(None)` and even `itertools.repeat(True, None)` both throw exceptions, rather than iterating forever. And I could pull this out into a helper, but think that whatever name I could pick would more more confusing than just inlineing this code.

Reply:
Looks like `islice` does the thing this needs! Just pushed a slightly cleaner commit.

Reply:
This block of code could use something like a big block comment at the top which explains what it is doing and why.  It's not clear to me what it's doing or why.

Also, `for _ in itertools.count()` will iterate forever...

Reply:
The code above doing the `islice` probably deserves a comment.  It wasn't clear to me on a quick read that it was also handling the possibility that `num_events` is `None` and thus needing it to do an infinite iteration in that case.

Reply:
Just pushed a version which is a more verbose but also easier to read.